sage_version_activity_to_csv.py

```diff
@@ -0,0 +1,318 @@
+            activity['rescanned'],
+            activity['latestNotableTimestamp'],
+            activity['notableCounts'] if len(activity['notableCounts']) > 0 else None
+    ]
+
+
+# -----------------------------------------------------------------------------
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(description="Process Sage project versions and check Hub for notable activity")
+
+    parser.add_argument("--token-file", dest='token_file', required=True, help="containing access token")
+    parser.add_argument("--no-verify", dest='verify', action='store_false', help="disable TLS certificate verification")
+    parser.add_argument('--timeout', dest='timeout', default=15.0, help="Connection timeout in seconds")
+    parser.add_argument('--retries', dest='retries', default=3, help="Maximum number of retries for a single request")
+    parser.add_argument('--skip-bom', dest='skip_bom', action='store_true', default=None, help="Skip BOM lookup")
+
+    group1 = parser.add_argument_group('required arguments')
+    group1.add_argument("--base-url", required=True, help="Hub server URL e.g. https://your.blackduck.url")
+    group1.add_argument('--input', dest='json_file_input', required=True, help="File containing Sage output e.g. sage_says.json")
+    group1.add_argument('--output', dest='csv_file_output', required=True, help="Output CSV file")
+
+    args = parser.parse_args()
+
+    if os.path.exists(args.csv_file_output) and os.path.samefile(args.json_file_input, args.csv_file_output):
+        print("Error, input and output file cannot be the same")
+        sys.exit(-1)
+
+    with open(args.json_file_input, 'r') as jf:
+        logging.info("Loading data from %s...", args.json_file_input)
+        sageJson = json.load(jf)
+        logging.info("Loaded data for %i codelocations across %i projects", len(sageJson['scans']), len(sageJson['projects']))
+
+    with open(args.token_file, 'r') as tf:
+        access_token = tf.readline().strip()
+
+    bd = Client(
+        base_url=args.base_url,
+        token=access_token,
+        verify=args.verify,
+        timeout=args.timeout,
+        retries=args.retries,
+    )
+
+    projectDict = {}   # key:projectId: json
+    pvDict = {}        # key:projectId: json array
+
+    for project in sageJson['projects']:
+        m = re.match(r".*/projects/(.*)", project['url'])
+        projectId = m.group(1)
+        projectDict[projectId] = project
+        pvDict[projectId] = project['versions']
+
+    codelocationsDict = {}  # key:codelocationId: json
+    for codelocation in sageJson['scans']:
+        m = re.match(r".*/codelocations/(.*)", codelocation['url'])
+        codelocationId = m.group(1)
+        codelocationsDict[codelocationId] = codelocation
+
+    # Process project versions
+    logging.info("Loading project versions complete, now processing each and every project version")
+
+    f = open(args.csv_file_output, 'w', newline='', encoding='utf-8')
+    w = csv.writer(f)
```

Comment:
DictWriter as opposed to just csv.writer? Not a big deal but just throwing it out there in case you find it easier to use DictWriter

Reply:
Ahh, good to know. I'll take a look at DictWriter.